tests/cache/test_cache.py

```diff
@@ -0,0 +1,239 @@
+        addr = 0x1234_5678
+        
+        # Initially state should be None (line not present)
+        self.assertIsNone(self.cache.get_line_state(addr))
+        
+        # Fill line with EXCLUSIVE state
+        self.cache.cache_line_fill(addr, MESIState.EXCLUSIVE)
+        
+        # Verify we can get the state
+        self.assertEqual(self.cache.get_line_state(addr), MESIState.EXCLUSIVE)
+        
+        # Verify we can set a new state
+        self.assertTrue(self.cache.set_line_state(addr, MESIState.MODIFIED))
+        self.assertEqual(self.cache.get_line_state(addr), MESIState.MODIFIED)
+        
+        # Verify setting state of non-existent line returns False
+        non_existent_addr = 0x5555_5555
+        self.assertFalse(self.cache.set_line_state(non_existent_addr, MESIState.MODIFIED))
+    
+    def test_snoop_line(self):
+        """Test snoop line lookup functionality"""
+        addr = 0x1234_5678
+        
+        # Initially line shouldn't be present
+        self.assertIsNone(self.cache.snoop_line(addr))
+        
+        # Add line in Modified state
+        self.cache.pr_write(addr)
+        self.cache.cache_line_fill(addr, MESIState.MODIFIED)
+        
+        # Set L1 inclusion status
+        addr_fields = self.cache._Cache__decompose_address(addr)
+        cache_set = self.cache.sets[addr_fields.index]
+        way_index = cache_set.find_way_by_tag(addr_fields.tag)
+        cache_set.ways[way_index].in_l1 = True
+        
+        # Check snoop results
+        line = self.cache.snoop_line(addr)
+        self.assertIsNotNone(line)
+        self.assertEqual(line.mesi_state, MESIState.MODIFIED)
+        self.assertTrue(line.in_l1)
+        
+        # Verify snoop doesn't affect cache state
+        self.assertEqual(
+            self.cache.get_line_state(addr),
+            MESIState.MODIFIED
+        )
+        
+        # Verify modifying returned line doesn't affect cache
+        line.mesi_state = MESIState.SHARED
+        self.assertEqual(
+            self.cache.get_line_state(addr),
+            MESIState.MODIFIED
+        )
+    
+    def test_cache_line_fill(self):
+        """Test cache line fill operation"""
+        addr = 0x1234_5678
+        
+        # Test initial fill
+        self.assertIsNone(self.cache.snoop_line(addr))  # Verify line not present
+        victim = self.cache.cache_line_fill(addr, MESIState.EXCLUSIVE)
+        self.assertIsNone(victim)  # No victim on initial fill
+        
+        # Verify line was properly filled
+        line = self.cache.snoop_line(addr)
+        self.assertIsNotNone(line)
+        self.assertEqual(line.mesi_state, MESIState.EXCLUSIVE)
+        
+        # Fill same line with different state
+        victim = self.cache.cache_line_fill(addr, MESIState.MODIFIED)
+        self.assertIsNone(victim)  # No victim when updating existing line
+        line = self.cache.snoop_line(addr)
+        self.assertEqual(line.mesi_state, MESIState.EXCLUSIVE)
```

Comment:
I'm a bit unclear on what this is testing, since `cache_line_fill` is set up to throw a warning indicating an existing line should not be updated by using `cache_line_fill`, right? If it's intended to test that the state is not updated, and that a warning is issued, we should maybe mock the warning call and check that it is called once. And then assigning values to something like `current_state` and `next_state` could maybe make it clearer in each assertion whether the line has the expected state given what's being tested.

Reply:
Correct, you are right. I should add more comments here or mock test. The main thing was to test that it would throw a warning. This was more for our purposes to help make sure we get operations ordered correctly.